### PR TITLE
HHH-14083  Add gradle task to automate the CI release

### PR DESCRIPTION
**Pull request by dreab8**
The PR add a new gradle task that automates the entire release process.
Running the command
`./gradlew clean ciRelease -PreleaseVersion=5.0.0.Final -PdevelopmentVersion=5.0.1-SNAPSHOT`
the following steps will be automatically executed :

1. the version is changed to `5.0.0.Final` (provided `releaseVersion`)
2.  the changelog.txt file is updated grabbing the Release Notes from Jira
3.  a git commit and push is executed
4.  the `relesase` task is kicked
5.  the release is tagged and the tag is pushed
6.  the version is changed to `5.0.1-SNAPSHOT` (provided `developmentVersion`)
7. a git commit and push is executed

the git remote used by default is `origin` the `gitRemote` property can be used to change the target remote.

the `noTag` parameter can be used to skip steps 5,6 and 7

Any suggestion is more than welcome :)

### CODE REVIEW

File: gradle/base-information.gradle [hunk 1]
```diff
@@ -7,8 +7,20 @@
 
 apply plugin: 'base'
 
+File versionFile = file( "${rootProject.projectDir}/gradle/version.properties" )
+
+if ( project.hasProperty( 'releaseVersion' ) ) {
+	if ( !project.hasProperty( 'developmentVersion' ) ) {
+		throw new GradleException(
+				"When releaseVersion parameter is specified also a developmentVersion have to be provided "
```
Comment (Sanne) [inline gradle/base-information.gradle:15]: I'd rephrase the error to english:
"When the `releaseVersion` parameter is specified, the parameter `developmentVersion` is required as well."

File: gradle/base-information.gradle [hunk 2]
```diff
@@ -7,8 +7,20 @@
...
+		)
+	}
+	versionFile.text = "hibernateVersion=${project.property( 'releaseVersion' )}"
+}
+
 ext {
-	ormVersion = new HibernateVersion( '5.5.0-SNAPSHOT', project )
+	ormVersionFile = versionFile
+	ormVersion = HibernateVersion.fromFile( versionFile, project )
```
Comment (Sanne) [inline gradle/base-information.gradle:23]: Damn this will conflict massively with my JPA3 WIP :/

Maybe one of us should wait for the other to finish?

File: release/release.gradle [hunk 3]
```diff
@@ -244,3 +246,170 @@ artifacts {
...
+        executeGitCommand(
+                "git push ${gitRemote}",
+                "An error occurred during the execution git push for the changelog.txt and the release version changes"
+        )
+    }
+}
+
+task ciRelease() {
+    group = "Release"
+    description = "Performs a release: the hibernate version is set and the changelog.txt file updated, the changes are pushed to github, then the release is performed, tagged and the hibernate version is set back to the development one"
```
Comment (Sanne) [inline release/release.gradle:279]: I'd remove "back" as it's not necessarily "going back" to the existing version. More likely to the next version?

File: release/release.gradle [hunk 4]
```diff
@@ -244,3 +246,170 @@ artifacts {
...
+    doLast {
+        if ( !project.hasProperty( 'developmentVersion' ) ) {
+            throw new GradleException(
+                    "When releaseVersion parameter is specified also a developmentVersion have to be provided "
```
Comment (Sanne) [inline release/release.gradle:283]: Rephrase (as above)

**Comment by dreab8** [comment]
Thanks @Sanne, I have applied your suggestions.

File: release/release.gradle [hunk 5]
```diff
@@ -244,3 +246,170 @@ artifacts {
...
+    proc.waitFor()
+    proc.consumeProcessErrorStream( errorsStringBuffer )
+    if ( errorsStringBuffer.toString(  ) != "" ) {
+        throw new GradleException( errorMessage + " " + b );
+    }
+}
+
+class ChangeLogFile {
+
+    // Get the Release Notes from Jira and add them to the Hibernate changelog.txt file
```
Comment (sebersole) [inline release/release.gradle:336]: It would be nice to clean up the blank lines the Jira report always seems to have.  This is something we normally do manually.

File: release/release.gradle [hunk 6]
```diff
@@ -242,5 +287,163 @@ artifacts {
...
+                int result = bis.read();
+                while (result != -1) {
+                    buf.write((byte) result);
+                    result = bis.read();
+                }
+                return buf.toString( StandardCharsets.UTF_8.name());
+            }
+        }
+    }
+}
```
Comment (NathanQingyangXu) [inline release/release.gradle:366]: could we replace the above method with `stream.getText("UTF-8")`?

File: release/release.gradle [hunk 7]
```diff
@@ -242,5 +287,163 @@ artifacts {
...
+                }
+        }
+        changelog.text = text
+    }
+
+    // Get the Release Notes from Jira
+    static String getNewReleaseNoteBlock(String releaseVersion) {
+        def restReleaseVersion;
+        if ( releaseVersion.endsWith( ".Final" ) ) {
+            restReleaseVersion = releaseVersion.replace( ".Final", "" )
```
Comment (NathanQingyangXu) [inline release/release.gradle:391]: The above code makes the assumption that `.Final` appears only once, which is a sound assumption. But
```
restReleaseVersion = releaseVersion.substring( 0, releaseVersion.lastIndexOf(".Final") )
```
would be more robust and accurate.

Comment (NathanQingyangXu) [reply release/release.gradle:391]: I think there is a perfect solution in Groovy in one-liner as
```
def restReleaseVersion = releaseVersion.replaceFirst( /\.Final$/, '' )
```
I tested it at https://groovyconsole.appspot.com/

**Comment by dreab8** [comment]
applied upstream
